[filebeat] add enhanced fingerprint to filestream (#50566)
elastic/beats: docs/reference/filebeat/filebeat-input-filestream.md

> However, in case of copytruncate strategy, you should provide additional configuration to Filebeat.
> 
> ::::{note}
> With the default `fingerprint` [file identity](/reference/filebeat/file-identity.md), you don't need `rotation.external.strategy.copytruncate` at all. The fingerprint follows the file content, so the input detects copytruncate rotation on its own. If you set it together with the `fingerprint` file identity, it is ignored.
> ::::

Reviewer: This is GA from 9.5+ right?

```suggestion
::::{note}
:applies_to: stack: ga 9.5+

With the default `fingerprint` [file identity](/reference/filebeat/file-identity.md), you don't need `rotation.external.strategy.copytruncate` at all. The fingerprint follows the file content, so the input detects copytruncate rotation on its own. If you set it together with the `fingerprint` file identity, it is ignored.
::::
```

> However, in case of copytruncate strategy, you should provide additional configuration to Filebeat.
> 
> ::::{note}
> With the default `fingerprint` [file identity](/reference/filebeat/file-identity.md), you don't need `rotation.external.strategy.copytruncate` at all. The fingerprint follows the file content, so the input detects copytruncate rotation on its own. If you set it together with the `fingerprint` file identity, it is ignored.
> ::::
> 

Reviewer 2: You're missing the applies_to tag:
`````suggestion
::::{note}
```{applies_to}
stack: preview 9.5+
```

With the default `fingerprint` [file identity](/reference/filebeat/file-identity.md), you don't need `rotation.external.strategy.copytruncate` at all. The fingerprint follows the file content, so the input detects copytruncate rotation on its own. If you set it together with the `fingerprint` file identity, it is ignored.
::::
`````

Reviewer 2: Maybe move this block after the "rotation.external.strategy.copytruncate" header.

Reviewer: On notes, the syntax for the `applies_to` is different, see [Admonitions > Applies to information](https://docs-v3-preview.elastic.dev/elastic/docs-builder/tree/main/syntax#applies-to-information).

Elsewhere the `applies_to` says GA, so if that’s the case, this should be:

````
::::{note}
:applies_to: stack: ga 9.5+

With the default `fingerprint` [file identity](/reference/filebeat/file-identity.md), you don't need `rotation.external.strategy.copytruncate` at all. The fingerprint follows the file content, so the input detects copytruncate rotation on its own. If you set it together with the `fingerprint` file identity, it is ignored.
::::
```

::::

> `fingerprint` is the default and recommended file identity because it does not rely on the file system/OS, it generates a hash from a portion of the file (the first 1024 bytes, by default) and uses that to identify the file. This works well with log rotation strategies that move/rename the file and on Windows as file identifiers might be more volatile. Before 9.5, the downside was that Filebeat waited until the file reached 1024 bytes before ingesting it; since 9.5 the [Enhanced Fingerprint](/reference/filebeat/file-identity.md#file-identity-fingerprint-growing) behavior (`file_identity.fingerprint.growing`, enabled by default) tracks files smaller than the fingerprint size, so they are ingested without delay.

Reviewer 2: Small style detail: I'd reverse the order of the last sentence so we first show the advantage of using Filebeat >= 9.5, then mention the drawback of using <9.5
```suggestion
`fingerprint` is the default and recommended file identity because it does not rely on the file system/OS, it generates a hash from a portion of the file (the first 1024 bytes, by default) and uses that to identify the file. This works well with log rotation strategies that move/rename the file and on Windows as file identifiers might be more volatile. Since 9.5 the [Enhanced Fingerprint](/reference/filebeat/file-identity.md#file-identity-fingerprint-growing) behavior (`file_identity.fingerprint.growing`, enabled by default) tracks files smaller than the fingerprint size, so they are ingested without delay; before 9.5, the downside was that Filebeat waited until the file reached 1024 bytes before ingesting it.
```

Or even edit it a bit more to start with something like "Since 9.5 files of any size can be ingested without delay..."

Reviewer 3: Done in c02d42b7a2

::::

> `fingerprint` is the default and recommended file identity because it does not rely on the file system/OS, it generates a hash from a portion of the file (the first 1024 bytes, by default) and uses that to identify the file. This works well with log rotation strategies that move/rename the file and on Windows as file identifiers might be more volatile. Since 9.5 the [Enhanced Fingerprint](/reference/filebeat/file-identity.md#file-identity-fingerprint-growing) behavior (`file_identity.fingerprint.growing`, enabled by default) tracks files smaller than the fingerprint size, so they are ingested without delay; before 9.5, the downside was that Filebeat waited until the file reached 1024 bytes before ingesting it.

Reviewer: ```suggestion
`fingerprint` is the default and recommended file identity because it does not rely on the file system/OS, it generates a hash from a portion of the file (the first 1024 bytes, by default) and uses that to identify the file. This works well with log rotation strategies that move/rename the file and on Windows as file identifiers might be more volatile.

{applies_to}`stack: ga 9.5+` The [Enhanced fingerprint](/reference/filebeat/file-identity.md#file-identity-fingerprint-growing) behavior (`file_identity.fingerprint.growing`, enabled by default) tracks files smaller than the fingerprint size, so they are ingested without delay. In earlier versions, Filebeat waits until the file reaches 1024 bytes before ingesting it.
```

file_identity.fingerprint: ~
```

The `fingerprint` file identity accepts a `growing` sub-option {applies_to}`stack: ga 9.5.0`. When `true` (the default since 9.5), files smaller than the fingerprint size (`offset`+`length`) are tracked using the bytes available so far instead of being skipped until they grow large enough, and are automatically migrated to the regular SHA-256 fingerprint once they reach the threshold, with no data duplication. Set it to `false` to restore the pre-9.5 behavior. See [Enhanced Fingerprint](/reference/filebeat/file-identity.md#file-identity-fingerprint-growing) for details.

```yaml
file_identity.fingerprint:
  growing: true

> ```

Reviewer: The applies_to should be at the front of the paragraph it relates to, or better yet, I think here we should use an indented sub-definition:

````suggestion
    `growing` {applies_to}`stack: ga 9.5.0+`
    :   When `true` (default), files smaller than the fingerprint size (`offset` + `length`) are tracked using the bytes available so far, instead of being skipped until they grow large enough. Once a file reaches the fingerprint size, it's automatically migrated to the regular SHA-256 fingerprint, with no data duplication.
    Refer to [Enhanced Fingerprint](/reference/filebeat/file-identity.md#file-identity-fingerprint-growing) for details.
    Set to `false` to restore the pre-9.5 behavior.

    ```yaml
    file_identity.fingerprint:
      growing: true
    ```
````

file_identity.fingerprint: ~
```

    `growing` {applies_to}`stack: ga 9.5.0+`
    :   When `true` (default), files smaller than the fingerprint size (`offset` + `length`) are tracked using the bytes available so far, instead of being skipped until they grow large enough. Once a file reaches the fingerprint size, it's automatically migrated to the regular SHA-256 fingerprint, with no data duplication.
    Refer to [Enhanced Fingerprint](/reference/filebeat/file-identity.md#file-identity-fingerprint-growing) for details.
    Set to `false` to restore the pre-9.5 behavior.

>     

Reviewer 2: I'm not sure the rendered version is very clear: 

<img width="1136" height="394" alt="Image" src="https://github.com/user-attachments/assets/e14b2bc2-5764-444e-b280-caa45e385319" />

I'd at least re add the example:
````
```yaml
file_identity.fingerprint:
  growing: true
```
````

Reviewer: I agree, I think I had the snippet in my original suggestion—maybe something in the suggestion was not quite right. 

The formatting is also off, I don’t think it renders as it should. I’ll check locally, and add a commit to fix this.

Reviewer: Ok, I’ve added a fix in https://github.com/elastic/beats/pull/50566/changes/d0dc9f4aa557e4f993a46fa613bf9bf43492736a (the sub-definition was indented a bit more than needed). I also re-added the snippet. Rendered:

<img width="764" height="791" alt="Screenshot 2026-07-03 at 7 08 35" src="https://github.com/user-attachments/assets/573ef874-08e6-46b5-a769-3d5cc4ea332e" />
